### DISCUSSION TITLE
Server Sent Events.

### DISCUSSION THREAD

**Discussion by chanpod** [sergiodxa/remix-utils#132] [Q&A] [answered]
I'm a little confused on the documentation for server sent events. You show this example

 let time = useEventSource("/sse/time", { event: "time" });
 
 But don't clarify that url. Is that url from which the event is being sent from or is it /sse/$eventName
 aka is /sse some route you create in remix to handle these events or was that just a demo route?

I've tried both but I don't receive anything in the UI.

Also does this work from actions inside a more complex object?

return {
...someData,
eventStream: eventStream(....)
}

**Comment by chanpod** [comment]
I'm trying to have a details page for an business object. And then on the edit page, when the object is updated, send out a SSE to hot update anyone looking at the details page. Looks like that's not quite how you have this setup (or is that even possible? I've never used SSE before to be honest).

**Comment by sergiodxa** [comment, marked as answer]
It's the route with the loader returning an `eventStream` response.

You can't return part of the data as an event stream, the event stream is a Response object similar to `json` or `redirect`, and it needs to be inside a loader because the browser does a GET request to connect.

**Reply by kael** [reply]
The doc is unclear that the:
- the loader for SSE must be used in a resource route ;
- the URL of the SSE route is the one of the resource route.

Naming the files in the code samples (with a specific comment for the resource route) would be nice.

**Comment by chanpod** [comment]
Ok that makes sense. Is it possible to send from another route? Or would the primary use case here be to setup a polling process on the server (querying for data for changes) and then sending updates as they happen?  Was hoping I could do a more reactive setup (user triggers an action and that action can send out events)

**Reply by sergiodxa** [reply]
The route from where you return the event stream is up to you, choose any pathname you want, just remember it needs to be a resource routes because UI routes can only return JSON or redirect responses.

And what you send and how you use the streamed data is up to your app, you could either send something and trigger a revalidation of the loader or send the data, keep it in a state and use it in the UI.

The `eventStream` and use `useEventSource` just helps you do the connection.

**Comment by vcheeze** [comment]
@sergiodxa thanks for the good work you are doing on remix-utils. I saw your blog post about a real-time chat app using SSE with Remix, but it doesn't seem to work for me when I use the latest Indie Stack and just add your code blocks. Specifically, what isn't working is the `emitter.on` handler not being caught. I'm new to Remix and am not entirely sure why attaching the handler in the route's loader function doesn't seem to work. I tried attaching the handler right after I instantiate the EventEmitter, and in that case the emitted event is caught and handled.

**Reply by vcheeze** [reply]
Update:

It is now working with the following code in my emitter:

```typescript
if (process.env.NODE_ENV === "production") {
  emitter = new EventEmitter();
} else {
  if (!global.__emitter) {
    global.__emitter = new EventEmitter();
  }
  emitter = global.__emitter;
}
```

It seems the dev environment won't recognize it as a global variable unless it is declared in the global scope.
